File: xv11test/examples/csharp/xv11pipe-loop/LaserReading.cs
/*
 * XV11 LIDAR test utility example
 *
 * Copyright (C) 2016 Bartosz Meglicki <meglickib@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 * This program is distributed "as is" WITHOUT ANY WARRANTY of any
 * kind, whether express or implied; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

using System;

namespace xv11pipeloop
{
    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential, Pack =1)]
    public struct LaserReading
    {

        /// distance : 14
        ///strength_warning : 1
        ///invalid_data : 1
        public ushort bitvector;

        /// unsigned short
        public ushort signal_strength;

        public ushort distance
        {
            get
            {
                return ((ushort)((this.bitvector & 16383u)));
            }
            set
            {
                this.bitvector = ((ushort)((value | this.bitvector)));
            }
        }

        public ushort strength_warning
        {
            get
            {
                return ((ushort)(((this.bitvector & 16384u)
                            / 16384)));
            }
            set
            {
                this.bitvector = ((ushort)(((value * 16384)
                            | this.bitvector)));
            }
        }

        public ushort invalid_data
        {
            get
            {
                return ((ushort)(((this.bitvector & 32768u)
                            / 32768)));
            }
            set
            {
                this.bitvector = ((ushort)(((value * 32768)
                            | this.bitvector)));
            }
        }
    }

    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential, Pack = 1)]
    public struct LaserFrame
    {

        /// BYTE->unsigned char
        public byte start;

        /// BYTE->unsigned char
        public byte index;

        /// unsigned short
        public ushort speed;

        /// laser_reading[4]
        [System.Runtime.InteropServices.MarshalAsAttribute(System.Runtime.InteropServices.UnmanagedType.ByValArray, SizeConst = 4, ArraySubType = System.Runtime.InteropServices.UnmanagedType.Struct)]
        public LaserReading[] readings;

        /// unsigned short
        public ushort checksum;
    }
}

File: xv11test/examples/java/src/com/bmegli/xv11lidar/LaserReading.java
package com.bmegli.xv11lidar;

/*
 * XV11 LIDAR test utility example
 *
 * Copyright (C) 2016 Bartosz Meglicki <meglickib@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 * This program is distributed "as is" WITHOUT ANY WARRANTY of any
 * kind, whether express or implied; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

import java.io.DataInput;
import java.io.IOException;

class LaserReading
{	
	private short bits;
	private short signal;
				
	public LaserReading() {
	}

	public void Fill(DataInput in) throws IOException
	{
		// bits and signal are in Little Endian so we can't use in.readUnsignedShort (Big Endian)
		bits = (short) ( in.readUnsignedByte() | ( in.readUnsignedByte() << 8) );
		signal =(short) ( in.readUnsignedByte() | ( in.readUnsignedByte() << 8) );		
	}
	
	public int getSignal()
	{
		return signal & 0xFFFF;
	}
	
	public int getDistance()
	{
		return bits & 0x3FFF;
	}
	
	public boolean isStrengthWarning()
	{
		return (bits & 0x4000) != 0;
	}
	
	public boolean isInvalidData()
	{
		return (bits & 0x8000) != 0;
	}	
}
